doc/api/stream.md

```diff
@@ -1867,6 +1873,57 @@ const myWritable = new Writable({
 });
 ```
 
+#### `writable._construct(options, callback)`
+<!-- YAML
+added: REPLACEME
+-->
+
+* `options` {Object} Options passed to the `Writable` constructor.
+* `callback` {Function} Call this function (optionally with an error
+  argument) when the stream has finished initializing.
+
+The `_construct()` method MUST NOT be called directly. It may be implemented
+by child classes, and if so, will be called by the internal `Writable`
+class methods only.
+
+This optional function will be called by the stream constructor,
+delaying any `_write`, `_final` and `_destroy` calls until `callback` is
+called. This is useful to initialize state or asynchronously initialize
+resources before the stream can be used.
+
+```js
+const { Writable } = require('stream');
+const fs = require('fs');
+
+class WriteStream extends Writable {
+  constructor(filename) {
+    super({ filename, autoDestroy: true });
+  }
+  _construct({ filename }, callback) {
+    this.filename = filename;
+    this.fd = null;
```

Comment:
This example feels a bit contrived, taking `filename` from `options` to demonstrate what that argument is. I suggest:

```js
constructor (filename) {
  super({ autoDestroy: true })
  this.filename = filename
  this.fd = null
}

_construct (options, callback) {
  fs.open(this.filename, ..)
}
```

Benefits: clear separation of builtin stream options vs userland options, and V8 likes it when you define properties in the constructor.

Makes me wonder why `_construct()` has an `options` argument at all?

Reply:
> Makes me wonder why _construct() has an options argument at all?

To avoid having to store essentially temporary properties on the instance.

Reply:
I agree, with this specific example it doesn't make much sense. Would make more sense if `filename` is not stored.

Reply:
> Benefits: clear separation of builtin stream options vs userland options, and V8 likes it when you define properties in the constructor.

Problem with your version is that `_construct` is called in `super` so you are essentially resetting the properties to `null`.

Reply:
> To avoid having to store essentially temporary properties on the instance.

That assumes such userland options are only needed in `_construct()`, which often isn't true. So I'm not sure there's a real benefit for use cases other than this specific example.

Reply:
@vweevers: If I understand correctly you would prefer no options then?

Reply:
@vweevers: I think I've resolved this according to your suggestions. PTAL

Reply:
Nice. I like this better, because the handling of options now only happens in (the function body of) the constructor, which leaves `_construct` with a sole concern: _asynchronous_ initialization.